### PR TITLE
Prepare 4.0.10 (4.0-edge)

### PR DESCRIPTION
**Pull request by simondeziel**

### CODE REVIEW

**Comment by simondeziel** [comment]
@tomponline please note that I included a zfs-2.1 bump from 2.1.6 to 2.1.15 as that's a LTS branch in upstream OpenZFS. If you feel that's too much changes, I'll remove it.

Also, for `dqlite` maybe we'd like to switch to the proper tag that was 2-3 commits after the one we had when 4.0.9 was built.

File: snapcraft.yaml [hunk 1]
```diff
@@ -704,6 +704,7 @@ parts:
       - edk2
       - qemu
     source: https://github.com/puiterwijk/qemu-ovmf-secureboot
+    source-commit: 4da81a4cbdd6d96b7a653239865d3b1d92cc6117 # master's HEAD
```
Comment (tomponline) [inline snapcraft.yaml:707]: At the time the stable snap was built?

Comment (simondeziel) [reply snapcraft.yaml:707]: Yes and also true to this date.

**Comment by simondeziel** [comment]
Doesn't build just yet as we'll need changes in `stable-4.0` due to `go.mod` referencing old module paths:

```
+ set -ex
+ cd ../src
+ git config user.email noreply@lists.canonical.com
+ git config user.name 'LXD snap builder'
++ realpath ./.go
+ export GOPATH=/root/parts/lxd/src/.go
+ GOPATH=/root/parts/lxd/src/.go
+ export 'CGO_CFLAGS=-I/root/stage/include/ -I/root/stage/usr/local/include/'
+ CGO_CFLAGS='-I/root/stage/include/ -I/root/stage/usr/local/include/'
+ export 'CGO_LDFLAGS=-L/root/stage/lib/ -L/root/stage/usr/local/lib/'
+ CGO_LDFLAGS='-L/root/stage/lib/ -L/root/stage/usr/local/lib/'
+ export 'CGO_LDFLAGS_ALLOW=(-Wl,-wrap,pthread_create)|(-Wl,-z,now)'
+ CGO_LDFLAGS_ALLOW='(-Wl,-wrap,pthread_create)|(-Wl,-z,now)'
+ go build -o /root/parts/lxd/install/bin/lxc github.com/canonical/lxd/lxc
no required module provides package github.com/canonical/lxd/lxc; to add it:
	go get github.com/canonical/lxd/lxc
Failed to run 'override-build': Exit code was 1.
Run the same command again with --debug to shell into the environment if you wish to introspect this failure.
```

**Comment by simondeziel** [comment]
@tomponline I've been able to build a snap using this diff (not part of the PR, ofc):

```
$ git show
commit 37dd11f876ce7c2b1889d0e107cc8729611c6173 (HEAD -> prepare-4.0.10-4.0-edge)
Author: Simon Deziel <simon.deziel@canonical.com>
Date:   2024-06-20 16:43:39 -0400

    snapcraft: test build with personal branch
    
    Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

diff --git a/snapcraft.yaml b/snapcraft.yaml
index 960118ea..b8adbc1a 100644
--- a/snapcraft.yaml
+++ b/snapcraft.yaml
@@ -1131,23 +1131,23 @@ parts:
       git config user.name "LXD snap builder"
 
       set +ex
       snapcraftctl build
       set -ex
 
       sed -i "s#\${LXC_ROOTFS_MOUNT}/var/snap/lxd/common/var/lib/lxcfs/#\${LXC_ROOTFS_MOUNT}/var/lib/lxcfs/#g" $SNAPCRAFT_PART_INSTALL/snap/lxd/current/lxcfs/lxc.mount.hook
       patch -p1 $SNAPCRAFT_PART_INSTALL/snap/lxd/current/lxcfs/lxc.mount.hook < "${SNAPCRAFT_PROJECT_DIR}/patches/lxcfs-0001-hook.patch"
 
   lxd:
-    source: https://github.com/canonical/lxd
+    source: https://github.com/simondeziel/lxd
     source-type: git
-    source-branch: stable-4.0
+    source-branch: prepare-4.0.10-stable-4.0
     after:
       - lxc
       - dqlite
       - sqlite
     build-packages:
       - libacl1-dev
       - libudev-dev
       - pkg-config
     build-snaps:
       - go/1.18/stable
```

I've used the resulting snap for some basic manual tests. The `images:` remote works fine. The only problem I found so far is that I'm unable to run nested VMs when the parent VM is 20.04+HWE. I'll try again with a 22.04 VM just to be 100% this is not a limitation with Focal. Could it be that 4.0 doesn't support being nested?

**Comment by simondeziel** [comment]
On a 22.04 VM (GA kernel) I get `lxc launch images:alpine/edge av1 --vm -c security.secureboot=false`:

```
# dmesg -c
[  167.544565] audit: type=1326 audit(1718920319.156:214): auid=4294967295 uid=0 gid=0 ses=4294967295 subj=lxd-av1_</var/snap/lxd/common/lxd> pid=4557 comm="qemu-system-x86" exe="/snap/lxd/x1/bin/qemu-system-x86_64" sig=31 arch=c000003e syscall=56 compat=0 ip=0x7f5a1393af3f code=0x80000000
```

**Comment by tomponline** [comment]
Does 4.0/stable work on 22.04?

**Comment by simondeziel** [comment]
> Does 4.0/stable work on 22.04?

Yes, so that confirms my build is at fault. Still investigating as a fresh/clean rebuild didn't help. Updating libseccomp to 2.5.5 didn't help either.

**Comment by simondeziel** [comment]
It turns out that I based my changes on top of `4.0-edge` but `4.0-candidate` apparently received an emergency cherry-pick (https://github.com/canonical/lxd/commit/af5c49e020877c238510eb096ac586e9708721a6, found in https://github.com/canonical/lxd-pkg-snap/compare/4.0-edge...4.0-candidate) due to a change made in glibc breaking our QEMU usage.

To avoid any cherry-pick applied in the `4.0-edge` branch of the `lxd-pkg-snap` repo I added that LXD commit to https://github.com/canonical/lxd/pull/13643 (which is a pre-requisite for this PR).

Doing the comparison the other way around shows that `4.0-edge` only adds support for `openvswitch.external`: https://github.com/canonical/lxd-pkg-snap/compare/4.0-candidate...4.0-edge

**Comment by simondeziel** [comment]
Basic manual tests are all passing now that we have the missing commit for VM support. I'm still trying to improve the QA coverage but this IMHO is ready to hit the `4.0/edge` channel (once the LXD required PR is merged as well).

**Review by tomponline — APPROVED** [review]
Thanks!

**Pull request merged**